keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,844 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Not all types following the Enum Pattern are actually enum types. Here is a counter-example.
+```go
+package core
+
+// StorageMedium defines ways that storage can be allocated to a volume.
+type StorageMedium string
+
+// These are the valid value for StorageMedium
+const (
+  StorageMediumDefault         StorageMedium = ""           // use whatever the default is for the node
+  StorageMediumMemory          StorageMedium = "Memory"     // use memory (tmpfs)
+  StorageMediumHugePages       StorageMedium = "HugePages"  // use hugepages
+  StorageMediumHugePagesPrefix StorageMedium = "HugePages-" // prefix for full medium notation HugePages-<size>
+)
+```
+
+Here, `StorageMedium` can have infinite number of possible values, which disqualify it as an enum type.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The syntax proposed in this KEP is different from existing implementation of kube-builder.
+Namely, kube-builder requires possible values in the marker itself, and is treated
+as a validation.
+```go
+package example
+
+// Alias
+// +kubebuilder:validation:Enum=Lion;Wolf;Dragon
+type Alias string
+```
+The two approaches are compatible, but eventually they will have to be unified.
+To retain compatibility with existing users of kube-builder, the auto-deduction will be added
+to kube-builder once the work is done with built-in types.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### OpenAPI v2 Breakage
+This feature, when enabled, breaks existing code using swagger to generate client or server stub from the spec
+for languages with built-in enum support, including Java and C#.
+Take the following spec as an example.
+
+```yaml
+definitions:
+  Foo:
+    type: object
+    properties:
+      bar:
+        type: string
+        enum: 
+         - A
+         - B
+         - C
+```
+
+The generated Java object looks like
+```java
+public class Foo {
+
+  @JsonAdapter(BarEnum.Adapter.class)
+  public enum BarEnum {
+    A("A"),
+    B("B"),
+    C("C");
+    // ...
+  }
+  private BarEnum bar = null;
+}
+```
+If the `enum` field was not in the definition, the `Bar` field would be a String.
+Maintainers of the clients should include the change in a new major release.
+
+If, for some reason, a client does not want enum fields to be generated, a simple preprocessor
```

Comment:
```suggestion
If, rather than update code generator or use of the generated client, a client would prefer enum fields to be strings as they were previously, a simple preprocessor
```

(or anything else indicating that this is a solution for the immediately preceding concern and not just a random paragraph :) )